### ISSUE TITLE
[RFC] Modularizing Sample Generation with Rating in PPO for Flexible RLHF Pipelines

### ISSUE THREAD

**Issue by zhuzilin** [OpenRLHF/OpenRLHF#499]
## TL; DR

This RFC proposes separating sample generation and reward model scoring from the original rollout process in PPO, enabling users to more flexibly customize sample generation and create samples with higher diversity.

## Current rollout procedure in PPO

1. Sample a batch of prompts with the size of `rollout_batch_size` from `prompt_dataset`;
2. Duplicate each prompt in the batch `n_samples_per_prompt` times, resulting in `n_samples_per_prompt * rollout_batch_size` prompts.
3. Group the `n_samples_per_prompt * rollout_batch_size`  prompts into micro-batches, each of size `micro_rollout_batch_size`;
4. For each micro batch:
  4.1. Generate responses with huggingface model or vllm;
  4.2. Rate the responses from the reward model or using `remote_reward_url`;
  4.3. Calculate `action_log_probs`, `kl` and `value` from actor, reference and critic models;
5. Save the results into a list of `Experience`;
6. Calculating advantages and returns;

In this structure, users have limited flexibility to dynamically adapt sample generation during training. For example, it is challenging to:
- Remove certain prompts during training if they are too difficult or too easy for the model.
- Balance the ratio of positive and negative responses for each prompt.
- Apply custom filtering to the samples, such as threshold-based filtering as shown in https://arxiv.org/pdf/2409.06957
- Create more complex sample generation procedures, e.g.,
  - prompting the model to self-refine or improve responses if initial ratings are low, as shown in https://arxiv.org/abs/2409.12917;
  - resampling responses with refined system prompts if all responses for a prompt have low ratings.

These limitations restrict the framework’s ability to create a diverse range of samples, hinder exploration, and reduce data efficiency in the training process.

## Proposed Change

This RFC proposes to modularize sample generation and rating (steps 4.1 and 4.2 in the current procedure) into a standalone component with the following functionality:

```python
def generate_samples_with_rating(prompt_dataset, rollout_size, vllm_engines, reward_models):
    """
    Generates and rates samples from a prompt dataset.
    
    Args:
        prompt_dataset (torch.utils.data.Dataset): The list of prompts to sample from.
        rollout_size (int): The number of prompts to select from the prompt dataset.
        vllm_engines (List): The list of language model engines (e.g., Hugging Face or vllm models) for generating responses.
        reward_models (List): The list of reward models for rating the responses.
    
    Returns:
        List[Tuple[List[int], float]]: A list of response and rating pair of size `rollout_size`
    """
    ...
```

This refactoring changes the rollout procedure to:

1. Get `rollout_size` (corresponding to the origin `n_samples_per_prompt * rollout_batch_size`) samples and its rating from `prompt_dataset` with vllm or hf model and the reward models;
2. Calculate `action_log_probs`, `kl` and `value` from actor and critic models;
3. Calculating advantages and returns;

This change allows users, especially those researching RLHF or O1, to more easily customize the sampling process. We expect models trained with OpenRLHF to achieve better results due to this flexibility.

## Engineering Benefits

This change has several engineering advantages:

- **Separation of Batch Sizes**: We can decouple the batch size used for prompts sent to vllm from the batch sizes used for the actor, reference, and critic models, maximizing sample generation performance. Additionally, it provides finer-grained control by allowing different batch sizes for the actor, reference, and critic models.
- **Optimized Communication**: Reduces overhead from inter-process communication by consolidating interactions between the main Ray process that handles rollouts and the Ray processes hosting the models.
- **Increased Scalability**: Allows an arbitrary number of vllm_engines, independent of the actor’s process group size.
- **Code Clarity**: Improves code readability and maintainability by isolating complex components.

A potential downside is that pipelining `action_log_probs`, `kl`, and `value` calculations with vllm generation becomes more challenging. However, attempting to pipeline these steps may result in convoluted code, undermining one of OpenRLHF’s strengths: clarity and readability.

Thank you for your time on this RFC. Looking forward to any discussion about it :)

**Comment by zhuzilin** [comment]
cc @hijkzzz

**Comment by hijkzzz** [comment]
The engineering advantages sound quite reasonable. The generation_batch_size for vLLM should ideally be a large value, whereas this may not necessarily be the case for Ref and Reward models, due to the matrix computations required by ZeRO3 parallelism. One point to consider is that Ray can inherently implement pipeline parallelism quite simply, so it might be possible to design an architecture compatible with this type of parallelism.

On the algorithm side, the advantages seem less convincing because PPO is an on-policy algorithm. Any behavior that filters or modifies samples violates the on-policy principle, effectively turning it into an off-policy algorithm.

**Comment by zhuzilin** [comment]
> Any behavior that filters or modifies samples violates the on-policy principle, effectively turning it into an off-policy algorithm. It can even lead to algorithm divergence.

Thank you for the feedback! I understand the concern regarding PPO’s on-policy nature. However, the proposed design primarily aims to offer flexibility, enabling researchers to experiment with different sampling strategies without enforcing off-policy behavior. Rather than introducing “training tricks” into the existing training protocol, this approach provides a flexible framework that users can adapt as needed.

> ```
> for bs0, bs1 and bs2
> 
>       vLLM     Reward+Ref
> t0    bs0
> t1    bs1          bs0        
> t2    bs2          bs1    
> t3                 bs2
> ```

Yes, this pipeline structure could serve as the default implementation of the proposed design—specifically, as a modular function for generation and rating in a pipeline fashion. However, it’s important to note that in the proposed structure, adding a third stage to calculate action_log_probs, kl, and value from the actor and critic models could be challenging. The main objective here is to separate the generation and rating from these other processes, allowing more modular control.

**Comment by hijkzzz** [comment]
We could start with a simplified version without pipelining, but it’s uncertain if it would be faster than the current implementation, as the current setup seems to achieve some degree of overlap in resource utilization. This is inherently achieved by Ray's scheduling mechanism.

**Comment by tsaoyu** [comment]
Thanks for the proposal. This provides an excellent starting point for implementing flexible training-time inference features, which will be crucial for scaling reinforcement learning in LLMs. I'd like to discuss several use cases that extend beyond MCTS.

## 1. Self-debate and Multi-agent Interaction

Consider a scenario with two agents ($$\pi_0$$ & $$\pi_1$$), where $$\pi_0$$ is the policy we aim to improve. The trajectory would look like:

$$ (s_0, a_{0}^{\pi_0}, a_{0}^{\pi_1}, a_{1}^{\pi_0}, a_{1}^{\pi_1}, \cdots, a_{n}^{\pi_0}, a_{n}^{\pi_1}) $$

The first challenge here is that this inference pattern may create a *cycle*, meaning our stage is no longer a DAG (Directed Acyclic Graph), even with clear termination conditions. This is particularly evident in self-debate scenarios where the next stage is self-referential.

## 2. Generative RM and Prompting-based Search

The second challenge arises when using generative reward models during inference time, which can break pipeline assumptions. Using $$\pi_0$$ to represent the policy and $$\pi_r$$ for the generative RM:

$$ (s_0, a_{0}^{\pi_0}, a_{0}^{\pi_r}, a_{1}^{\pi_0}, a_{1}^{\pi_r}, \cdots, a_{n}^{\pi_0}) $$

In this pattern, the reward model is called within the inference loop. This means we cannot always assume there's a pipelining opportunity between policy inference and reward model inference.

Similarly, prompting-based search (as seen in DSPy) demonstrates how reward models might be called at any point in the loop to encourage exploration.

## 3. Value-guided Generation

The final case involves using value functions to guide tree expansion in LLM generation. The key challenge here lies in effectively passing different components while maintaining OpenRLHF's simplicity.

## Implementation Considerations

While these patterns aren't exhaustive, they should help us explore our research limitations. Here are several implementation options based on my experience:

### 1. Reusing Well-defined Agentic Workflows (Flexibility)

My first approach was to leverage existing agentic languages. I chose LangGraph because its API design effectively mimics `networkx` and shares concepts with Google's `beam`. However, this comes with tradeoffs:

- **Downside**: The complex structure intertwined with LangChain makes it unnecessarily heavy
- **Potential**: LangGraph could become the "Flink" of AI, handling batching and streaming optimization

From a dataflow perspective, our core challenge is achieving flexible streaming while maintaining strong batching performance during training-time inference.

I've also explored DSPy, which excels in LLM expressiveness and prompting techniques. However, as an in-context learning framework where training only occurs in content, it may be overkill for our needs.

Currently, I'm considering something between plain text concatenation and OpenAI's `swarm`. The appeal lies in its simplicity – it's essentially plain text concatenation with basic agent interaction:

```python
from swarm import Swarm, Agent

client = Swarm()

def transfer_to_agent_b():
    return agent_b

agent_a = Agent(
    name="Agent A",
    instructions="You are a helpful agent.",
    functions=[transfer_to_agent_b],
)

agent_b = Agent(
    name="Agent B",
    instructions="Only speak in Haikus.",
)

response = client.run(
    agent=agent_a,
    messages=[{"role": "user", "content": "I want to talk to agent B."}],
)
```

For this approach, I suggest keeping the implementation lightweight and text-centric. Whether O1 is single or multi-agent, we can ensure our framework remains flexible enough to accommodate different inference patterns.

(Note: We should also consider interruptible decoding patterns for function calls or completed reasoning steps.)

### 2. Performance Optimization

Once we have a working `client`, we should focus on maximizing inference throughput during training, as this will be the bottleneck for any self-play RL algorithm. Based on my experience optimizing LangGraph-based inference workflows:

- Implement lazy evaluation
- Enable opportunistic batching
- Leverage existing optimizations from `vLLM` or `SGLang`

Prefill and decoding separation appears promising for long O1 inference, but we need to abstract it properly.

### 3. Flow Readability

While LangGraph offers flexibility, its readability suffers due to separated node definitions and edge connections. I suggest:

- Extracting mini-flows from the main self-play training loop
- Enabling independent debugging of inference flows
- Using a "plug-and-play" approach once the inference pipeline is validated

Thoughts on this approach?
